### PR TITLE
Add in auto-recovery for all operations

### PR DESCRIPTION
**Pull request by michaeldauria**

### CODE REVIEW

File: lib/promiscuous/amqp/rubyamqp.rb [hunk 1]
```diff
@@ -21,8 +21,19 @@ def self.connect
         end
 
         connection = ::AMQP.connect(amqp_options)
-        self.channel = ::AMQP::Channel.new(connection)
+        self.channel = ::AMQP::Channel.new(connection, :auto_recovery => true)
         self.queue_options = Promiscuous::Config.queue_options
+
+        connection.on_connection_interruption do |connection|
+          connection.periodically_reconnect(2)
+        end
```
Comment (michaeldauria) [inline lib/promiscuous/amqp/rubyamqp.rb:28]: this is when the connection is just broken as in a hard failure

File: lib/promiscuous/amqp/rubyamqp.rb [hunk 2]
```diff
@@ -21,8 +21,19 @@ def self.connect
...
+
+        connection.on_error do |connection, connection_close|
+          # CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'
+          if connection_close.reply_code == 320
+            connection.periodically_reconnect(2)
+          end
+        end
```
Comment (michaeldauria) [inline lib/promiscuous/amqp/rubyamqp.rb:35]: this is when we issue `service rabbitmq-server stop`

**Comment by nviennot** [comment]
What about `connection.on_tcp_connection_loss` ? if you kill the TCP connection in your test does it survive ?

**Comment by nviennot** [comment]
Any chance we can have an integration test also ? (because ruby-amqp can be upgraded, and we don't want to die)

**Comment by kareemk** [comment]
Integration test is a bitch, we'll have to add starting/stopping of rabbit to the responsibility of the suite (like https://github.com/ruby-amqp/amqp/blob/master/spec/integration/recovery/per_channel_automatic_recovery_on_graceful_broker_shutdown_spec.rb#L41-58)

And https://github.com/ruby-amqp/amqp/blob/master/spec/spec_helper.rb

**Pull request closed without merging**